site/src/pages/RoadmapPage/YearModal.tsx

```diff
@@ -0,0 +1,154 @@
+import React, { FC, useState } from 'react';
+import { Button, Form, Modal } from 'react-bootstrap';
+import { PlannerYearData } from '../../types/types';
+import './YearModal.scss';
+
+interface YearPopupQuarter {
+  id: string;
+  name: string;
+  checked?: boolean;
+}
+
+interface CourseYearModalProps {
+  placeholderName: string;
+  placeholderYear: number;
+  show: boolean;
+  setShow: React.Dispatch<React.SetStateAction<boolean>>;
+  type: 'add' | 'edit';
+  saveHandler: (x: PlannerYearData) => void;
+  currentQuarters: string[];
+}
+
+const CourseYearModal: FC<CourseYearModalProps> = (props) => {
+  const { placeholderName, placeholderYear, show, setShow, type, saveHandler, currentQuarters } = props;
+  const [validated, setValidated] = useState(false);
+
+  const [name, setName] = useState(placeholderName);
+  const [year, setYear] = useState(placeholderYear);
+
+  const quarterValues: (selectedNames: string[]) => YearPopupQuarter[] = (data: string[]) => {
+    const base: YearPopupQuarter[] = [
+      { id: 'fall', name: 'Fall' },
+      { id: 'winter', name: 'Winter' },
+      { id: 'spring', name: 'Spring' },
+      { id: 'summer 1', name: 'Summer 1' },
+      { id: 'summer 2', name: 'Summer 2' },
+      { id: 'summer 10 Week', name: 'Summer 10 Week' },
+    ];
+    data.forEach((name) => {
+      const match = base.find((q) => q.id === name);
+      if (match) match.checked = true;
+    });
+    return base;
+  };
```

Comment:
This can be moved outside the YearModal functional component. Slight optimization since it will avoid redefining it every render. Also, why not use a record object here mapping id to { name, checked } like:
```ts
{
  'fall': { name: 'Fall' },
  'winter': { name: 'Winter' },
  'spring': { name: 'Spring' },
  'summer 1': { name: 'Summer 1' },
  'summer 2': { name: 'Summer 2' },
  'summer 10 Week': { name: 'Summer 10 Week' },
}
```
Then you can just update it by indexing for each quarter id instead of using `find`.

Reply:
We need access to the keys, and the order matters. I should be able to extract it outside of the component though

Reply:
Ok. `Object.keys` and `Object.entries` could be used but I suppose that's no better than the current implementation.